### PR TITLE
Don't manually replace whitespace during render

### LINKED ISSUE #10156: gitea change spaces (%20) to signal +
- Gitea version (or commit ref): Latest version in try.gitea.io
- Git version:
- Operating system:
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [X] Yes (provide example URL) 
    Navigate in this url https://try.gitea.io/heliomcp/testes/src/branch/master/Teste%20A between the 2 files **using markdown links**.
  - [ ] No
  - [ ] Not relevant
- Log gist:

## Description

Every time I have a directory name with a space and try to use markdown links, gitea change the space to '+' instead of '%20' and it gives 404.
...

## Screenshots

**Original Markdown**
image

**Render OK**
image

**BAD Link** (directory with space) 
image

### PR DESCRIPTION
**Pull request by mrsdizzie**
For historical reasons Gitea manually alters the urlPrefix and replaces a whitespace with a ```+```. This Works for URLs, but we're also passing urlPrefix to git calls and adding the ```+``` is breaking the tree path.

Goldmark will automatically convert a white space to the proper ```%20```, so we should leave the string as is which lets us pass it to git unmodified and then let Goldmark fix any whitespaces when rendering. This keeps treePath and the generated link working. See #10156 for more details.

Fixes #10156

Here is example of broken + behavior that is fixed:

 https://try.gitea.io/mrsdizzie/testcase/src/branch/master/test%20a

### CODE REVIEW

File: modules/util/url.go [hunk 1]
```diff
@@ -25,6 +25,9 @@ func PathEscapeSegments(path string) string {
 
 // URLJoin joins url components, like path.Join, but preserving contents
 func URLJoin(base string, elems ...string) string {
+	// We do need to escape special chars here or else they can be silently discarded
+	// in the ResolveReference call below
+	base = PathEscapeSegments(base)
```
Comment (guillep2k) [inline modules/util/url.go]: If `base` requires escaping of its components, maybe `elems` need that too?

Comment (mrsdizzie) [reply modules/util/url.go]: Good catch, fixed

**Comment by mrsdizzie** [comment]
Hmm will need more though on escaping elems here

**Comment by mrsdizzie** [comment]
@guillep2k had to revert escaping the elements because it breaks valid fragment links like mypage#section1 and would involve more logic rewriting in general. I notice a few other bugs with filenames that have a pound sign/how we handle what we think are anchors in markdown and I think all of that logic needs to be looked into in a separate PR.

This PR will still fix issues mentioned above with regards to folder names and filenames with white spaces with just a few lines.

**Comment by guillep2k** [comment]
> 
> This PR will still fix issues mentioned above with regards to folder names and filenames with white spaces with just a few lines.

OK. Please send an empty commit to restart CI.

**Comment by guillep2k** [comment]
I've been reading the code, and now I'm not so sure about `URLJoin` escaping the `base` parameter. The function is used everywhere, and some places might have the base already escaped (e.g. what happens with utf-8 domain names? Or a Gitea URL that has a relative path?). Perhaps it's a better idea to have a `GitPathToHttpUrl()` function, and it when necessary, leaving `URLJoin` to do only one job.

**Comment by mrsdizzie** [comment]
> I've been reading the code, and now I'm not so sure about `URLJoin` escaping the `base` parameter. The function is used everywhere, and some places might have the base already escaped (e.g. what happens with utf-8 domain names? Or a Gitea URL that has a relative path?). Perhaps it's a better idea to have a `GitPathToHttpUrl()` function, and it when necessary, leaving `URLJoin` to do only one job.

You're right, anything that needs to should be escaped before being passed to that function or needs more thought anyways. 

I've just remove that change from this PR as it should be part of another refactor that addresses #10300 better (which is how I came across it). Updated PR description as well.

**Review by guillep2k — APPROVED** [review]

**Review by zeripath — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Pull request merged**